js/single-docset-search.tsx

```diff
@@ -308,16 +349,57 @@ const search = (
   searchIndex: Index | undefined,
   query: string
 ): SearchResult[] | undefined => {
-  const results =
-    searchIndex &&
-    searchIndex.search(query, {
-      bool: "OR",
-      expand: true,
-      fields: {
-        name: { boost: 5 },
-        doc: { boost: 2 }
+  if (!searchIndex) {
+    return;
+  }
+
+  // we'd like to favour exact matches, ex: clojure.tools.test
+  const exactTokens = tokenize(query);
+  // but also entertain sub tokens, ex: clojure tools test
+  const subTokens = subTokenize(exactTokens);
+
+  const fieldQueries = [
+    { field: "name", boost: 10, tokens: exactTokens },
+    { field: "name", boost: 5, tokens: subTokens },
+    { field: "doc", boost: 4, tokens: exactTokens },
+    { field: "doc", boost: 2, tokens: subTokens }
+  ];
+
+  const queryResults: elasticlunr.SearchScores = {};
+
+  for (const fieldQuery of fieldQueries) {
+    const searchConfig = {
+      [fieldQuery.field]: { boost: fieldQuery.boost, bool: "OR", expand: true }
+    };
+
+    const fieldSearchResults: elasticlunr.SearchScores =
+      searchIndex.fieldSearch(
+        fieldQuery.tokens,
+        fieldQuery.field as keyof IndexItem,
+        searchConfig
+      );
+
+    for (const docRef in fieldSearchResults) {
+      fieldSearchResults[docRef] =
+        fieldSearchResults[docRef] * fieldQuery.boost;
+    }
+
+    for (const docRef in fieldSearchResults) {
+      if (docRef in queryResults) {
+        queryResults[docRef] += fieldSearchResults[docRef];
+      } else {
+        queryResults[docRef] = fieldSearchResults[docRef];
       }
-    });
+    }
+  }
+  const results = [];
+  for (const docRef in queryResults) {
+    results.push({ ref: docRef, score: queryResults[docRef] });
+  }
```

Comment:
```typescript
const results = queryResults.map(docRef => ({ ref: docRef, score: queryResults[docRef] }));
```

Reply:
ya, that's leveling up!

Reply:
Actually... `docRef in queryResult` means docRef is an array index, no?
Perhaps it is not too bad to leave this one as is.

Reply:
Gonna leave this one.

Reply:
Ahhh right, the `in`, it's such a tricky difference. Great catch!